File: 2846-robot-collisions/robot-collisions.cpp
class Solution {
public:
    vector<int> survivedRobotsHealths(vector<int>& positions, vector<int>& healths, string directions) {
        vector<tuple<int,int,char,int>> r;
        for (int i = 0; i < healths.size(); i++) {
            r.push_back({positions[i], healths[i], directions[i], i});
        }
        sort(r.begin(), r.end());

        stack<tuple<int,int,char,int>> st;
        vector<int> res(healths.size(), 0);

        for (auto &[pos, h, dir, idx] : r) {
            if (st.empty() || dir == 'R') {
                st.push({pos, h, dir, idx});
            } else {
                while (!st.empty() && dir == 'L' && get<2>(st.top()) == 'R') {
                    auto [p, h1, d, index] = st.top();
                    if (h > h1) {
                        st.pop();
                        h--;
                    } else if (h == h1) {
                        st.pop();
                        h = 0;
                        break;
                    } else { 
                        st.pop();
                        st.push({p, h1 - 1, d, index});
                        h = 0;
                        break;
                    }
                }
                if (h > 0) {
                    st.push({pos, h, dir, idx});
                }
            }
        }

        while (!st.empty()) {
            auto [p, h1, d, idx] = st.top();
            res[idx] = h1;
            st.pop();
        }

        vector<int> ans;
        for (auto it : res) {
            if (it != 0) ans.push_back(it);
        }

        return ans;
    }
};

File: 2846-robot-collisions/robot-collisions.py
class Solution:
    def survivedRobotsHealths(self, positions: List[int], healths: List[int], directions: str) -> List[int]:
        
        arr=[]
        n=len(positions)
        for i in range(n):
            arr.append((positions[i],healths[i],directions[i],i))
        arr.sort(key=lambda x:x[0])
        st=[]
        for pos,health,direction,idx in arr:
            if len(st)==0:
                st.append((pos,health,direction,idx))
                continue
            while st and st[-1][2]=='R' and direction=='L':
                p,h,d,i=st[-1]
                if h<health:
                    st.pop()
                    health-=1
                    
                    
                elif h==health:
                    health=0
                    st.pop()
                    break
                    
                else:
                    health=0
                    st.pop()
                    st.append((p,h-1,d,i))
                    break
            if health>0:
                st.append((pos,health,direction,idx))
        
            
            

        st.sort(key=lambda x:x[3])
        ans=[]
        for pos,health,direction,idx in st:
            ans.append(health)
        return ans

            
        